pandas/core/internals.py

```diff
@@ -589,13 +589,16 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
                               "CategoricalDtype instead",
                               FutureWarning, stacklevel=7)
 
-            kwargs = kwargs.copy()
-            categories = getattr(dtype, 'categories', None)
-            ordered = getattr(dtype, 'ordered', False)
+            categories = kwargs.get('categories', None)
+            ordered = kwargs.get('ordered', None)
+            if _any_not_none(categories, ordered):
+                dtype = CategoricalDtype(categories, ordered)
 
-            kwargs.setdefault('categories', categories)
-            kwargs.setdefault('ordered', ordered)
-            return self.make_block(Categorical(self.values, **kwargs))
+            if is_categorical_dtype(self.values):
```

Comment:
I don't think you need all of this logic, wouldn't

```
values = self.values.astype(dtype, copy=copy)

return self.make_block(values, dtype=dtype)
```
be enough (if values is a Categorical already or dtype is a CDT, it will infer correctly, and if its not it will as well).

Reply:
I don't think that quite works, since `self.values` can be a different object depending on what `self` is: if `self` is already categorical, then `self.values` is a `Categorical`, otherwise `self.values` is a numpy array.

In the numpy case, `self.values.astype` raises `TypeError: data type not understood` when a `CDT` is passed as the dtype.

Likewise, `self.make_block(Categorical(self.values, dtype=dtype))` also doesn't work by itself.  In the `Categorical` case, the constructor ignores the `dtype` parameter when the input data is already `Categorical`, so no update occurs.

Seems like the two paths are necessary?  Or am I overlooking something?

Reply:
ok seems reasonable then